File: app/src/main/java/com/example/rafaelanastacioalves/moby/domain/entities/Resource.java
package com.example.rafaelanastacioalves.moby.domain.entities;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

public class Resource<T> {
    @NonNull  public Status status;
    @Nullable public T data;
    @Nullable public String message;

    private Resource(@NonNull Status status, @Nullable T data, @Nullable String message){
        this.status = status;
        this.data = data;
        this.message = message;
    }

    public static <T> Resource<T> success(@NonNull T data){
        return new Resource<>(Status.SUCCESS, data, null);
    }

    public static <T> Resource<T> error(Status status, @Nullable T data, String msg){
        return new Resource<>(status, data, msg );
    }

    public static <T> Resource<T> loading(@Nullable T data){
        return new Resource<>(Status.LOADING, data, null);
    }

    public enum Status { SUCCESS, INTERNAL_SERVER_ERROR ,GENERIC_ERROR, LOADING}
}

File: app/src/main/java/com/example/rafaelanastacioalves/moby/domain/entities/Resource.kt
package com.example.rafaelanastacioalves.moby.domain.entities

class Resource<T> constructor(
        status: Status,
        val data: T?,
        var message: String?) {

    companion object Factory {

        fun <T> success(successData: T?): Resource<T>  {
            return Resource(Status.SUCCESS,successData, null);
        }

        fun <T> error(status: Status, data: T?, msg: String?): Resource<T> {
            return Resource(status, data, msg)
        }

        fun <T> loading(): Resource<T> {
            return Resource(Status.LOADING, null, null)
        }
    }

    enum class Status { SUCCESS, INTERNAL_SERVER_ERROR ,GENERIC_ERROR, LOADING}
}